pwiz_tools/Skyline/Controls/FilesTree/FileSystemHealthMonitor.cs

```diff
@@ -0,0 +1,213 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Threading;
+using pwiz.Common.SystemUtil;
+using pwiz.Skyline.Util.Extensions;
+
+namespace pwiz.Skyline.Controls.FilesTree
+{
+    internal class FileSystemHealthMonitor : IDisposable
+    {
+        private readonly TimeSpan _checkInterval;
+        private ManualResetEvent _manualTriggerEvent = new ManualResetEvent(false);
+        private readonly object _lock = new object();
+
+        private Thread _workerThread;
+        private CancellationTokenSource _cancellationTokenSource;
+        private Dictionary<string, PathAvailability> _paths;
+
+        private bool _isWorking;
+
+        internal event Action<string, PathAvailability> PathAvailabilityChanged;
+
+        public FileSystemHealthMonitor(TimeSpan checkInterval)
+        {
+            _checkInterval = checkInterval;
+            _paths = new Dictionary<string, PathAvailability>();
+        }
+
+        public void Start()
+        {
+            // If already started, stop the existing thread first to prevent multiple threads
+            // from accessing the same _paths and _manualTriggerEvent
+            if (_workerThread != null)
+            {
+                Stop();
+            }
+
+            // Reset stopping/stopped flags so AddPath() can be called again
+            _isStopped = _isStopping = false;
+
+            _cancellationTokenSource = new CancellationTokenSource();
+
+            // ActionUtil.RunAsync() already starts the thread, so don't call Start() again
+            _workerThread = ActionUtil.RunAsync(() => MonitorLoop(_cancellationTokenSource.Token),
+                @"FilesTree => FileSystemAvailabilityMonitor");
+        }
+
+        public void AddPath(string newPath)
+        {
+            // Enforce usage sequence: Start() -> AddPath() -> Stop()
+            // AddPath() should not be called after Stop() has been called
+            Assume.IsFalse(_isStopping || IsStopped(),
+                @"AddPath() called after Stop(). Expected sequence: Start() -> AddPath() -> Stop()");
+
+            lock (_lock)
+            {
+                if (!_paths.ContainsKey(newPath))
+                {
+                    _paths[newPath] = PathAvailability.unknown;
+                }
+            }
+        }
+
+        private void MonitorLoop(CancellationToken cancellationToken)
+        {
+            while (!cancellationToken.IsCancellationRequested)
+            {
+                if (_isWorking)
+                {
+                    SpinWait.SpinUntil(() => !cancellationToken.IsCancellationRequested, 200);
```

Comment:
The "!" looks incorrect in this line of code (and on line 122 as well).
As it is, this code will wait 200msec if cancelled or 0msec if not cancelled.

Reply:
Cleaned this area up quite a bit. Thanks for the pointer.